WordPressKit/WordPressComOAuthClient.swift

```diff
@@ -1,15 +1,106 @@
+    /// WordPress.com returns an HTTP response that WordPressKit can't parse.
+    case unparsableResponse(response: HTTPURLResponse?, body: Data?)
+    /// Other error occured.
+    case unknown(underlyingError: Error)
+
+    public var authenticationFailureKind: AuthenticationFailure.Kind? {
+        if case let .authenticationFailure(authenticationFailure) = self {
+            return authenticationFailure.kind
+        }
+        return nil
+    }
+
+    public struct AuthenticationFailure {
+        private static let errorsMap: [String: WordPressComOAuthError.AuthenticationFailure.Kind] = [
+            "invalid_client": .invalidClient,
+            "unsupported_grant_type": .unsupportedGrantType,
+            "invalid_request": .invalidRequest,
+            "needs_2fa": .needsMultifactorCode,
+            "invalid_otp": .invalidOneTimePassword,
+            "user_exists": .socialLoginExistingUserUnconnected,
+            "invalid_two_step_code": .invalidTwoStepCode,
+            "unknown_user": .unknownUser
+        ]
+
+        public enum Kind {
+            /// client_id is missing or wrong, it shouldn't happen
+            case invalidClient
+            /// client_id doesn't support password grants
+            case unsupportedGrantType
+            /// A required field is missing/malformed
+            case invalidRequest
+            /// Multifactor Authentication code is required
+            case needsMultifactorCode
+            /// Supplied MFA code is incorrect
+            case invalidOneTimePassword
+            /// Returned by the social login endpoint if a wpcom user is found, but not connected to a social service.
+            case socialLoginExistingUserUnconnected
+            /// /// Supplied MFA code is incorrect
+            case invalidTwoStepCode
+            case unknownUser
+            case unknown
+        }
+
+        public var kind: Kind
+        public var localizedErrorMessage: String?
+        public var newNonce: String?
+        public var originalErrorJSON: [String: AnyObject]
+
+        init(apiJSONResponse responseDict: [String: AnyObject]) {
+            originalErrorJSON = responseDict
+
+            // there's either a data object, or an error.
+            if  let errorStr = responseDict["error"] as? String {
+                kind = Self.errorsMap[errorStr] ?? .unknown
+                localizedErrorMessage = responseDict["error_description"] as? String
+            } else if let data = responseDict["data"] as? [String: AnyObject],
+                let errors = data["errors"] as? NSArray,
+                let err = errors.firstObject as? [String: AnyObject] {
+                let errorCode = err["code"] as? String ?? ""
+                kind = Self.errorsMap[errorCode] ?? .unknown
+                localizedErrorMessage = err["message"] as? String
+                newNonce = data["two_step_nonce"] as? String
+            } else {
+                kind = .unknown
+            }
```

Comment:
Have you considered using `Decodable` support types for this? Or... would it be too convoluted because of hte nested dictionaries?